test/extensions/clusters/common/logical_host_integration_test.cc

```diff
@@ -10,60 +10,47 @@ namespace Extensions {
 namespace Clusters {
 
 // Logical Host integration test.
-class LogicalHostIntegrationTest : public testing::TestWithParam<Network::Address::IpVersion>,
-                                   public HttpIntegrationTest {
+class LogicalHostIntegrationTest : public testing::Test, public HttpIntegrationTest {
 public:
   LogicalHostIntegrationTest()
-      : HttpIntegrationTest(Http::CodecType::HTTP1, GetParam()),
+      : HttpIntegrationTest(Http::CodecType::HTTP1, Network::Address::IpVersion::v6),
```

Comment:
I agree that the crux of the problem is getting the internal Envoy threads (Envoy's + dns-resolver?) under control.
I'm not aware of anything that the test-server supports in this case, and looking at [this comment](https://github.com/envoyproxy/envoy/blob/f580ed067db0f2a48085aa33e45a3a9cf5e36bfe/test/integration/base_integration_test.h#L486-L489) I'm guessing that more subtle support may be needed in test-server if a generic solution is needed.
I wonder if the mocked method be removed, or better yet can the dns resolver in the Envoy be overridden (returned to default state)? I think that should take care of Envoy calling the resolver's method.

Reply:
Can we try to use the server lifecycle notifier to signal an absl::Notification and block on that notification from the test thread before exiting?

Reply:
@jmarantz  Let's continue discuss whether there is a way to get the Envoy internal threads under control. 

BTW, if "static" is not preferred, we can use std::rand() to generate a random number here, which will practically cause the address list be different with large probability to satisfy this check:
https://github.com/envoyproxy/envoy/blob/1ae6fb29cab8043cc23e070b118fa25c79f7f7d8/source/extensions/clusters/logical_dns/logical_dns_cluster.cc#L136
That way, we can still reproduce the original issue https://github.com/envoyproxy/envoy/issues/32850 with below test command:

bazel test -c dbg test/extensions/clusters/common:logical_host_integration_test  --config=remote --config=clang-tsan --runs_per_test=100

Reply:
@adisuissa 
I tried to shut down the mock DNS resolving by something like below at the end of the test:
@@ -97,6 +97,8 @@
//------diff------- 
   ASSERT_TRUE(response->complete());
   EXPECT_EQ("200", response->headers().getStatusValue());
+  // Stop DNS resolving.
+  EXPECT_CALL(*dns_resolver, resolve(_, _, _)).WillRepeatedly(testing::Return(nullptr));
 }
 
 } // namespace Clusters

This can shutdown DNS resolving thus avoid the MSAN issue. However, this changes bring in TSAN issue as the DNS thread is reading the mocker while the test thread is writing it thus abandoned this approach.

Reply:
I tihnk we need to clean up all state including threads in between test methods.